CA1062 false positive when switching on (obj?.Property)

@mavasani Is there a way to find out if https://github.com/dotnet/roslyn-analyzers/pull/2584 already fixes this?

No diagnostic should be reported in this method:

```cs
public static void Foo(Task task)
{
    switch (task?.Status)
    {
        // ...

        case TaskStatus.Faulted:
            //  ↓ CA1062 In externally visible method 'void Foo(Task task)', validate parameter 'task' is non-null before using it. If appropriate, throw an ArgumentNullException when the argument is null or add a Code Contract precondition asserting non-null argument.
            _ = task.Exception;
            break;

        // ...
    }
}
```

#### Analyzer package
https://www.nuget.org/packages/Microsoft.CodeQuality.Analyzers/

#### Package Version

2.9.3

#### Diagnostic ID

https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1062-validate-arguments-of-public-methods

I'm also getting this from `dotnet build` using SDK 3.0.100-preview3-010431 at the same source location:

> warning CS8602: Possible dereference of a null reference.

Maybe this is a Roslyn bug.

--- a/src/Microsoft.CodeQuality.Analyzers/UnitTests/QualityGuidelines/ValidateArgumentsOfPublicMethodsTests.cs
+++ b/src/Microsoft.CodeQuality.Analyzers/UnitTests/QualityGuidelines/ValidateArgumentsOfPublicMethodsTests.cs
@@ -5732,5 +5732,66 @@ public sealed class B : A
     protected override bool IsType(Type type) => type.Namespace == nameof(System);
 }");
         }
+
+        [Trait(Traits.DataflowAnalysis, Traits.Dataflow.NullAnalysis)]
+        [Fact, WorkItem(2526, "https://github.com/dotnet/roslyn-analyzers/issues/2526")]
+        public void CheckedWithConditionalAccess_01()
+        {
+            VerifyCSharp(@"
+using System.Collections.Generic;
+
+public class C
+{
+    public bool Flag;
+    public void M(C c)
+    {
+        if (c?.Flag == true)
+        {
+          var x = c.ToString();
+        }
+    }
+}");
+        }
+
+        [Trait(Traits.DataflowAnalysis, Traits.Dataflow.NullAnalysis)]
+        [Fact, WorkItem(2526, "https://github.com/dotnet/roslyn-analyzers/issues/2526")]
+        public void CheckedWithConditionalAccess_02()
+        {
+            VerifyCSharp(@"
+using System.Collections.Generic;
+
+public class C
+{
+    public bool M(List<string> list)
+    {
+        return list?.Count > 5 && list.Count < 10;
+    }
+}");
+        }
+
+        [Trait(Traits.DataflowAnalysis, Traits.Dataflow.NullAnalysis)]
+        [Fact, WorkItem(2586, "https://github.com/dotnet/roslyn-analyzers/issues/2586")]
+        public void CheckedWithConditionalAccess_03()
+        {
+            VerifyCSharp(@"
+using System.Collections.Generic;
+
+public class C
+{
+    public bool Flag;
+    public void M(C c)
+    {
+        switch (c?.Flag)
+        {
+            case true:
+                var x = c.ToString();
+                break;
+            case null:
+                var y = c.ToString();
+                break;
+        }
+    }
+}");
+        }
     }
 }

--- a/src/Utilities/Compiler/Extensions/ISymbolExtensions.cs
+++ b/src/Utilities/Compiler/Extensions/ISymbolExtensions.cs
@@ -546,11 +546,11 @@ namespace Analyzer.Utilities.Extensions
                     return ((IParameterSymbol)symbol).Type;
 
                 default:
-                    return GetMemerType(symbol);
+                    return GetMemberType(symbol);
             }
         }
 
-        public static ITypeSymbol GetMemerType(this ISymbol symbol)
+        public static ITypeSymbol GetMemberType(this ISymbol symbol)
         {
             switch (symbol.Kind)
             {

--- a/src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/ParameterValidationAnalysis/ParameterValidationAnalysis.ParameterValidationDataFlowOperationVisitor.cs
+++ b/src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/ParameterValidationAnalysis/ParameterValidationAnalysis.ParameterValidationDataFlowOperationVisitor.cs
@@ -426,12 +426,9 @@ namespace Microsoft.CodeAnalysis.FlowAnalysis.DataFlow.ParameterValidationAnalys
             {
                 var value = base.VisitIsNull(operation, argument);
 
-                // Mark a location as validated on paths where we know it is non-null.
+                // Mark a location as validated on paths where user has performed an IsNull check.
                 // See comments in VisitBinaryOperatorCore override above for further details.
-                if (FlowBranchConditionKind == ControlFlowConditionKind.WhenFalse)
-                {
-                    MarkValidatedLocations(operation.Operand);
-                }
+                MarkValidatedLocations(operation.Operand);
 
                 return value;
             }

--- a/src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/PointsToAnalysis/PointsToAnalysis.PointsToDataFlowOperationVisitor.cs
+++ b/src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/PointsToAnalysis/PointsToAnalysis.PointsToDataFlowOperationVisitor.cs
@@ -927,7 +927,7 @@ namespace Microsoft.CodeAnalysis.FlowAnalysis.DataFlow.PointsToAnalysis
                     symbol.Name.Equals("Empty", StringComparison.Ordinal) &&
                     (symbol.IsReadOnlyFieldOrProperty() || symbol.Kind == SymbolKind.Method) &&
                     (symbol.ContainingType.SpecialType != SpecialType.None ||
-                     symbol.GetMemerType() is INamedTypeSymbol namedType &&
+                     symbol.GetMemberType() is INamedTypeSymbol namedType &&
                      symbol.ContainingType.DerivesFromOrImplementsAnyConstructionOf(namedType.OriginalDefinition));
             }
 

